### ISSUE TITLE
How to train the model (tl_ssd) with our own dataset?

### ISSUE THREAD

**Issue by windyrainy** [julimueller/tl_ssd#9]

**Comment by julimueller** [comment]
We did not upload scripts to train it. In general, you can use the training scripts of the original repository and add the necessary parameters needed for our adaptations in these layers.

**Comment by EmilCreatePro** [comment]
Have you succeeded in training the model with the DTLD dataset? 
I am trying to convert the data to LMDB for some time, but with no success.. :(

I saw @julimueller that you mentioned using the adapted scripts from weiliu89/caffe,
For LMDB creation you used (and adapted?) the create_list.sh and create_data.sh? And for training you used the ssd_pascal.py script mentioned in that repository?

**Comment by julimueller** [comment]
@EmilCreatePro I adapted this script - it is not too complicated. Labels are then in the format

classid_1 xmin_1 ymin_1 xmax_1 ymax_1 
classid_2 xmin_2 ymin_2 xmax_2 ymax_2
....

so nothing too special. I then adapted the multiboxloss layer to extract the state class from the class identity. You only have to consider that in DTLD off has the digit 0 and this may lead to problems because the standard background label in the wiliu implementation is also zero. So it makes sense to change the state IDs before.

Initially I started with the ssd_pascal.py script, yes. But with time I almost changed everything to get a cleaner state. In the meantime my training pipeline blew up quite a lot.

**Comment by EmilCreatePro** [comment]
@julimueller I guess that all the adaption is to give the path to the DTLD dataset folders and change from .jpg type to .tiff

But running the create_list.sh script always gives me error:
_get_image_size.cpp:110] This tool requires OpenCV; compile with USE_OPENCV._

I have opencv 4.2 and 3.2 installed on my system, therefore I didn't set USE_OPENCV = 1 flag when compiling caffe(from what I read if I have it installed I don't need the flag on true). 

**If I compile with that flag I get many errors** when building caffe with a lot of functions being undefined ( like cv::imread(cv::String const&, int), cv::VideoCapture::open(int), cv::String::deallocate() etc... ) and don't know why. 

When you compiled you set the USE_OPENCV flag to true? It can't work without it?

I would be happy If I could just run the training and then try to tweak some hyperparameters.

**Comment by julimueller** [comment]
Well I created a small python script which loads the dataset and then stores all DTLD images in one folder and creates the required label txt files in another folder. So I never applied the script on the original DTLD folder structure...

Why dont you just compile ssd with this flag? cmake .. -DUSE_OPENCV=1 && make install should do the job then. I guess you should do this even though opencv is  installed. It will not work without and it should not take you much time...

**Comment by EmilCreatePro** [comment]
Ah, so the dataset handling **cannot** be done by any of the existence scripts. **And how is this creation of the required label txt files done?** I read in the documentation about Class Identity in the DTLD. Should the python script parse the .yml files and extract the **class_id** minimum and maximum of each object? I don't quite understand who those _xmin_1 ymin_1 xmax_1 ymax_1_ are supposed to be.

I tried your suggestions and _cmake .. -DUSE_OPENCV=1_ gives me
_CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDA_cublas_device_LIBRARY (ADVANCED)_

Didn't find CUDA_cublas_device_LIBRARY in the CMakeLists.txt and not even in the whole caffe-ssd folder.. 

I build the ssd using _make all -j12_ according to weilliu instructions. If I compile with USE_OPENCV flag set on 1 in the Makefile then I get these errors:
![TerminalSS](https://user-images.githubusercontent.com/30869379/80538841-0edf0380-89af-11ea-9af6-8809aaeb88b0.png)

I tried to patch up the sources files, but then same errors appear for other source files and is a nightmare to fix and I don't think its the proper way. Have you encountered something like this? Or do you have some advice?

**Comment by julimueller** [comment]
There are various pages about how to create detection lmdb files. See here: https://xiaoyuliu.github.io/2018/02/23/combine-datasets-into-lmdb-for-caffe/

I do not believe you did not find cublas in the repository.

![image](https://user-images.githubusercontent.com/36820065/80567946-53da5880-89f6-11ea-9c87-54ecfb15a268.png)

I do not use Makefile but cmake build. Your error messages above look like you still did not correctly link against opencv ...
